### ISSUE TITLE
getting the ELBO loss as a PyTorch Variable

### ISSUE THREAD

**Issue by fehiepsi** [uber/pyro#583]
I am trying to get the ELBO loss as a PyTorch Variable, not a float value. Checking Pyro's source code, I think that `surrogate_loss_particle` in `Trace_ELBO` class is what I want. Is there an easy way to obtain it, without mimic `Trace_ELBO.py` source code? Thank you very much in advance!

**Comment by martinjankowiak** [comment]
@fehiepsi what is your use case? the reason this is returned as a float is because in the case of N particles we don't want the computational graph to growth with N (which is what would happen if we kept around N non-detached PyTorch ELBO Variables)

**Comment by fehiepsi** [comment]
@martinjankowiak I am implementing Laplace approximation (just for self-studying how Pyro works and comparing to other packages such as PyMC3 or Edward). It quires calculate MAP and hessian of posterior at the mean/mode (they are same for Gaussian distribution). Hessian of posterior is the same to hessian of joint distribution. But joint distribution is also ELBO if we define guide as delta distribution for the latent variable `z` (I am not confident with this claim, I think in Pyro, the entropy of delta function is zero). Currently, PyTorch supports higher order derivatives, so getting "exact" Hessian using PyTorch is not a problem. I just need to get the loss, to apply `torch.autograd.grad` (which is used to compute higher order derivatives) to this loss. Here is an example, to calculate hessian matrices of sum of outputs w.r.t. each parameter `x` (1d Variables):
```python
def hessians(ys, xs):
    y = ys.sum()
    if isinstance(xs, Variable):
        xs = (xs,)
    for x in xs:
        assert x.dim() == 1
    y_xs = torch.autograd.grad(y, xs, create_graph=True)
    hess = tuple(torch.stack([torch.autograd.grad(y_x[i], x, retain_graph=True)[0].data
                             for i in range(x.size(0))])
                 for x, y_x in zip(xs, y_xs))
    if len(hess) == 1:
        hess = hess[0]
    return hess

**Comment by fritzo** [comment]
@fehiepsi I agree with @martinjankowiak that our current setup is preferable because it uses less memory. If memory were no issue, you could implement this by refactoring `SVI.loss_and_grads()` to return a `(loss, surrogate_loss)` pair rather than returning `loss` and calling `surrogate_loss.backward()` internally.

Can you think of a way to implement a streaming-accumulator version of `hessians()` that could be passed into `.loss_and_grads()` as a callback to be called once per particle (in place of `torch_backward()`)? That would achieve both of our design goals: low memory footprint + flexibility.

**Comment by eb8680** [comment]
@fehiepsi thanks for doing this, it's helpful for finding the rough edges in our API.  A further problem with implementing Laplace approximation right now is that we removed our multivariate normal distribution because some gradients are missing from PyTorch.

@fritzo I'm not sure that will work - this issue is interacting weirdly with the problem of automatically generating guides for MAP.  The cleanest way to implement Laplace approximation is probably to run SVI with a delta guide, compute the Hessian using `hessians`, then transform the delta guide into a bunch of Gaussians with the correct covariances.  To do that last step, we'd need a `TransformPoutine` primitive that modifies the distributions and input arguments at sample sites.  This would be useful elsewhere e.g. for transforming random variables from constrained to unconstrained support.

**Comment by fehiepsi** [comment]
@fritzo Thanks for your suggestions! Here is what comes to my mind (not rigorous). Please correct me if I understand things wrong:
```
from pyro.infer.map import MAP

def hessians(ys, xs):
    ...

class Laplace(MAP):
    def get_hessians():
        xs = ... # get parameters to calculate hessians
        # callback is a tuple with first element is a function, and remaining elements is some of its arguments
        loss, hs = self.loss_and_grads(self.model, self.guide, callback=(hessians, xs), *args, **kwargs)
        h = accumulate(hs)
        # hs is a list of "tuples of hessians"
        # len(hs) = len(particles)
        # len(hs[0]) = len(xs)
        return h

# modify Trace_ELBO
class Trace_ELBO():
    def loss_and_grads(self, model, guide, callback=None, , *args, **kwargs):
        cb_returns = []
        for each particle:
            ...
            cb_return = torch_backward(surrogate_loss_particle, callback)
            cb_returns.append(cb_return)
        if callback is None:
            return loss
        else:
            return loss, cb_returns

# modify torch_backward
def torch_backward(loss, callback=None):
    cb_return = None
    if callback is not None:
        cb_return = callback[0](loss, *callback[1:])
    loss.backward()
    return cb_return
```
I think it can achieve my purpose (I just need Normal guide or Diagonal Gaussian guide, not the complicated case which requires matrix's inverse: Multivariate Gaussian guide).

@eb8680's approach is really nice (thank you!) It fits my intuition of what are the results of MAP and Laplace inferences (a delta guide and a Gaussian guide). However, it is beyond my current understanding about Pyro to make it work.

**Comment by ngoodman** [comment]
@fritzo's solution and @eb8680's fancier solution are good directions... 

my 2c, i wonder whether a "by hand" first implementation might be a good idea? in particular, you could make a copy of the trace_elbo implementation, and change it to keep everything as Variables (and retain the graph, etc.). using this, first do MAP with delta guide to get the rv vals and hessions you need, then make a gaussian guide for the laplace approx. if this all works, making it more automatic would be swell.

more generally, @fehiepsi, i love the addition of laplace approx. looking forward to playing with it!

**Comment by fehiepsi** [comment]
@ngoodman I made a simple implementation at pr #590 (I don't know which is a better way to deliver it).  Now, I am thinking about making Gaussian guide. Assume that the name of Variable which we want to obtain is `x`, is it true that we will make a guide with parameters `x.mean` and `x.cov`, then sample `x` from these parameters?

**Comment by fehiepsi** [comment]
@fritzo Could Pyro support a `callback_on_loss` arg in `loss` or `loss_and_grads` functions? For example
```
def loss_and_grads(...):
    callback = kwargs.pop("callback_on_loss", None)   # we can replace None by torch_backward here
    cb_returns = []
    for each_trace:
        loss = ...
        cb_return = callback(loss)
        if cb_return is not None:
             cb_returns.append(cb_return)
    return (loss, cb_returns) if cb_returns else loss
```
I think that this additional arg will not affect much the current state of Pyro. The `callback_on_loss` (or better name) will be used at user's risk. In fact, in the `loss_and_grads`, `callback` can be `torch_backward` as default.

I wish to have this feature to be able to calculate Hessians, though it is sad that I can't think of any other use case. :( Maybe it is useful to observe the derivative of loss w.r.t. some particular variables.

**Comment by fritzo** [comment]
@fehiepsi I think the cleanest way to implement what you want would be to refactor `TraceGraph_ELBO()` a bit by moving the `.backwards()` invocations from `._loss_and_grads_particle()` downstream into `.loss_and_grads()`. Then you could implement a new `TraceGraph_ELBO.laplace_approx()` method or sth that calls `hessians()` on the `surrogate_loss` Variables returned by `_loss_and_grads_particle()`. Would that serve your purpose, or am I missing something?

**Comment by fehiepsi** [comment]
@fritzo Yes, it perfectly serves my purpose. This way would simplify the code to attain my purpose a lot.

**Comment by yhalk** [comment]
Hi,

If we can't keep track of the loss gradients due to memory consumption reasons, is there a way to monitor their evolution in order to inspect the training progress?

Thank you in advance for any hint!

**Comment by martinjankowiak** [comment]
@yhalk you can do something like follows
```
from pyro.infer.trace_elbo import Trace_ELBO

elbo = Trace_ELBO()
# since the following line populates grad attributes, you may want
# to zero out grads after this call, depending on what you're doing
elbo.loss_and_grads(model, guide)  
```
after this code snippet, the various trainable parameters in the paramstore have their `.grad` attributes populated, which you're now free to inspect. note that this won't take an actual gradient step. if you want to do that you could e.g. reproduce the logic in the `step` method of the `SVI` class

**Comment by fritzo** [comment]
@yhalk that's a great question. I've added some [gradient monitoring code](https://github.com/uber/pyro/blob/df2d171de1ce70848b0e1c92624fe783c428deb6/tutorial/source/gmm.ipynb) to the GMM tutorial in #907. I hope that helps.

**Comment by yhalk** [comment]
@fritzo @martinjankowiak Thank you both for your answers - I greatly appreciate the help!
 
Regarding the gradient monitoring code in #907 why do you use norm().item() ? I am getting a Variable there and had to use .data to access the gradient value.

Thanks again!

**Comment by jpchen** [comment]
> I am getting a Variable there and had to use .data to access the gradient value.

in pytorch 0.4 (master) `.item()` serves the same purpose as `.data()`

**Comment by yhalk** [comment]
@jpchen Thank you!

**Comment by fehiepsi** [comment]
Hi @jpchen, can we reopen this? The topic's question is still not yet solved. When we have MAP inference, I will make a pull request to address it (together with Laplace inference) following @fritzo's suggestion at https://github.com/uber/pyro/issues/583#issuecomment-348339390

**Comment by fehiepsi** [comment]
Closed because the feature raised by this issue is no longer necessary. Thanks for all discussion!
